File: Kotlin/55.jump_game.kt
fun main() {
    println(canJump(intArrayOf(2,3,1,1,4))) // true
    println(canJump(intArrayOf(3,2,1,0,4))) // false
    println(canJump(intArrayOf(2, 0))); // true
    println(canJump(intArrayOf(2, 5, 0, 0))); // true
    println(canJump(intArrayOf(0, 1))); // false
}

fun canJump(nums: IntArray): Boolean {
    var reachable = 0;

    for(i in nums.indices) {
        if (i > reachable) {
            return false;
          }
      
          reachable = Math.max(reachable, i + nums[i]);
    }
  
    return true;
}

File: lib/55.jump_game.dart
import 'dart:math';

void main() {
  print(canJump([2, 3, 1, 1, 4])); // true
  print(canJump([3, 2, 1, 0, 4])); // false
  print(canJump([2, 0])); // true
  print(canJump([2, 5, 0, 0])); // true
  print(canJump([0, 1])); // false
}

/// Time Limit Exceeded
// bool canJump(List<int> nums) {
//   var index = 0;

//   while (index < nums.length - 1) {
//     if (nums[index] == 0) {
//       break;
//     }

//     if (index + nums[index] < nums.length - 1) {
//       var max = 0;
//       var index2 = 0;

//       for (int i = index + nums[index]; i > index; i--) {
//         if (nums[i] > max) {
//           max = nums[i];
//           index2 = i;
//         }
//       }

//       index += index2;
//     } else {
//       index += nums[index];
//     }
//   }

//   return index >= nums.length - 1;
// }

bool canJump(List<int> nums) {
  var reachable = 0;

  for (int i = 0; i < nums.length; i++) {
    if (i > reachable) {
      return false;
    }

    reachable = max(reachable, i + nums[i]);
  }

  return true;
}
